pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/providerconfig_test.go

```diff
@@ -132,37 +139,139 @@ var _ = Describe("Provider Config", func() {
 		)
 	})
 
+	Context("NewProviderConfigFromMachine", func() {
+		type providerConfigTableInput struct {
+			modifyMachine         func(tmpl *machinev1beta1.Machine)
+			providerSpecBuilder   resourcebuilder.RawExtensionBuilder
+			providerConfigMatcher types.GomegaMatcher
+			expectedPlatformType  configv1.PlatformType
+			expectedError         error
+		}
+
+		DescribeTable("should extract the config", func(in providerConfigTableInput) {
+			machine := resourcebuilder.Machine().WithProviderSpecBuilder(in.providerSpecBuilder).Build()
+
+			if in.modifyMachine != nil {
+				in.modifyMachine(machine)
+			}
+
+			providerConfig, err := NewProviderConfigFromMachine(*machine)
+			if in.expectedError != nil {
+				Expect(err).To(MatchError(in.expectedError))
+				return
+			}
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(providerConfig.Type()).To(Equal(in.expectedPlatformType))
+			Expect(providerConfig).To(in.providerConfigMatcher)
+		},
+			Entry("with an invalid platform type", providerConfigTableInput{
+				modifyMachine: func(in *machinev1beta1.Machine) {
+					var awsProviderConfig machinev1beta1.AWSMachineProviderConfig
+					err := json.Unmarshal(in.Spec.ProviderSpec.Value.Raw, &awsProviderConfig)
+					Expect(err).To(BeNil())
+
+					awsProviderConfig.TypeMeta.Kind = "InvalidProviderSpecKind"
+					in.Spec.ProviderSpec.Value.Raw, err = json.Marshal(awsProviderConfig)
+					Expect(err).To(BeNil())
+				},
+				providerSpecBuilder: resourcebuilder.AWSProviderSpec(),
+				expectedError:       fmt.Errorf("could not determine platform type: %w", fmt.Errorf("%w: %s", errUnknownProviderConfigType, "InvalidProviderSpecKind")),
+			}),
+			Entry("with an AWS config with failure domains", providerConfigTableInput{
+				expectedPlatformType:  configv1.AWSPlatformType,
+				providerSpecBuilder:   resourcebuilder.AWSProviderSpec(),
+				providerConfigMatcher: HaveField("AWS().Config()", *resourcebuilder.AWSProviderSpec().Build()),
+			}),
+		)
+	})
+
+	Context("ExtractFailureDomainsFromMachines", func() {
+
+		type extractFailureDomainsFromMachinesTableInput struct {
+			machines               []machinev1beta1.Machine
+			expectedError          error
+			expectedFailureDomains []failuredomain.FailureDomain
+		}
+
+		awsSubnet := machinev1.AWSResourceReference{
+			Type: machinev1.AWSFiltersReferenceType,
+			Filters: &[]machinev1.AWSResourceFilter{
+				{
+					Name: "tag:Name",
+					Values: []string{
+						"aws-subnet-12345678",
+					},
+				}}}
```

Comment:
I'm generally opposed to closing multiple brackets like this, it looks weird and is hard to follow, better to use the indentation top align
```suggestion
				},
			},
		}
```